From af51133e202a328ba5d9ed532dd7145ce034b062 Mon Sep 17 00:00:00 2001
From: Zahari Doychev <zdoychev@maxlinear.com>
Date: Tue, 16 Mar 2021 10:42:53 +0100
Subject: [PATCH] dp: arp_ndp_reinsert: don't free skb on xmit error

PONRTSYS-8526 the dp_xmit frees always the skb so free it only in
case of LGM when busy error code is returned.
---
 drivers/net/datapath/bridge_filter/arp_ndp_reinsert.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/net/datapath/bridge_filter/arp_ndp_reinsert.c b/drivers/net/datapath/bridge_filter/arp_ndp_reinsert.c
index c93a69a7a640..a464c08f65b2 100644
--- a/drivers/net/datapath/bridge_filter/arp_ndp_reinsert.c
+++ b/drivers/net/datapath/bridge_filter/arp_ndp_reinsert.c
@@ -65,7 +65,9 @@ static int arp_ndp_reinsert_packet(struct sk_buff *skb)
 
 	ret = dp_xmit(skb->dev, &subif, nskb, nskb->len, DP_TX_INSERT);
 	if (ret) {
-		dev_kfree_skb_any(nskb);
+		/* DPM can return busy on LGM then skb needs to be freed */
+		if (ret == DP_TX_FN_BUSY)
+			dev_kfree_skb_any(nskb);
 		netdev_err(skb->dev, "%s: insert fail\n", __func__);
 		return -EIO;
 	}
